### PR TITLE
remove `megatron.training` dependency from `megatron.core` for FSDP checkpoint with EP

### PR DESCRIPTION
**Pull request by ananthsub**
# What does this PR do ?
Make num_experts be explicitly passed for checkpointing support instead of trying to read from `megatron.training` global args, which aren't guaranteed to be initialized when using `megatron.core`

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [x] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/fsdp_dtensor_checkpoint.py [hunk 1]
```diff
@@ -450,6 +475,7 @@ def get_global_unique_param_name(model_chunks, param):
             param_name = re.sub(r"layers\.(\d+)", f"layers.{tf_layer_number - 1}", param_name)
 
     # Get EP unique parameter name
-    param_name = list(handle_experts_in_state_dict({param_name: None}).keys())[0]
+    num_experts = model_chunks[0].config.num_moe_experts if model_chunks else None
+    param_name = list(handle_experts_in_state_dict({param_name: None}, num_experts).keys())[0]
```
Comment (Skylion007) [inline megatron/core/transformer/fsdp_dtensor_checkpoint.py]: You can use next)iter) for const time access to the first key.

File: megatron/core/optimizer/distrib_optimizer.py [hunk 2]
```diff
@@ -1153,7 +1153,8 @@ def _param_name(self, param: torch.nn.Parameter) -> str:
                         "Ensure that each model chunk has unique parameter names."
                     )
                 name_to_param.update(_name_to_param)
-            name_to_param = handle_experts_in_state_dict(name_to_param)
+            num_experts = self.model_chunks[0].config.num_moe_experts if self.model_chunks else None
```
Comment (yaoyu-33) [inline megatron/core/optimizer/distrib_optimizer.py]: use unwrap_model api?

File: megatron/core/transformer/fsdp_dtensor_checkpoint.py [hunk 3]
```diff
@@ -450,6 +475,7 @@ def get_global_unique_param_name(model_chunks, param):
             param_name = re.sub(r"layers\.(\d+)", f"layers.{tf_layer_number - 1}", param_name)
 
     # Get EP unique parameter name
-    param_name = list(handle_experts_in_state_dict({param_name: None}).keys())[0]
+    num_experts = model_chunks[0].config.num_moe_experts if model_chunks else None
```
Comment (yaoyu-33) [inline megatron/core/transformer/fsdp_dtensor_checkpoint.py]: use unwrap_model?

File: megatron/core/transformer/fsdp_dtensor_checkpoint.py [hunk 4]
```diff
@@ -47,29 +47,35 @@
...
     ep_rank = parallel_state.get_expert_model_parallel_rank()
-    num_local_experts = args.num_experts // ep_size if args.num_experts else 0
+    num_local_experts = num_experts // ep_size if num_experts else 0
     local_expert_offset = ep_rank * num_local_experts
 
     return local_expert_offset
 
 
-def get_total_num_experts():
+def get_total_num_experts(num_experts: int | None = None) -> int:
```
Comment (jaredcasper) [inline megatron/core/transformer/fsdp_dtensor_checkpoint.py]: This function feels totally unnecessary now, just inline it below.

**Review by yaoyu-33 — APPROVED** [review]
